src/k8s/pkg/k8sd/api/certs_refresh.go

```diff
@@ -0,0 +1,277 @@
+	)
+	if err != nil {
+		return response.InternalError(err)
+	}
+
+	csrKubeletClient, pKeyKubeletClient, err := pkiutil.GenerateCSR(
+		pkix.Name{
+			CommonName:   fmt.Sprintf("system:node:%s", snap.Hostname()),
+			Organization: []string{"system:nodes"},
+		},
+		2048,
+		nil,
+		nil,
+		nil,
+	)
+
+	_, err = client.CreateCertificateSigningRequest(
+		s.Context,
+		fmt.Sprintf("k8sd-%d-worker-kubelet-client", seed),
+		[]byte(csrKubeletClient),
+		[]v1.KeyUsage{v1.UsageDigitalSignature, v1.UsageKeyEncipherment, v1.UsageClientAuth},
+		nil,
+		"k8sd.io/kubelet-client",
+	)
+	if err != nil {
+		return response.InternalError(err)
+	}
+
+	csrKubeProxy, pKeyKubeProxy, err := pkiutil.GenerateCSR(
+		pkix.Name{
+			CommonName: "system:kube-proxy",
+		},
+		2048,
+		nil,
+		nil,
+		nil,
+	)
+
+	_, err = client.CreateCertificateSigningRequest(
+		s.Context,
+		fmt.Sprintf("k8sd-%d-worker-kube-proxy-client", seed),
+		[]byte(csrKubeProxy),
+		[]v1.KeyUsage{v1.UsageDigitalSignature, v1.UsageKeyEncipherment, v1.UsageClientAuth},
+		nil,
+		"k8sd.io/kube-proxy-client",
+	)
+
+	result := apiv1.RefreshCertificatesPlanResponse{
+		Seed: int(seed),
+		CertificatesSigningRequests: []string{
+			fmt.Sprintf("k8sd-%d-worker-kubelet-serving", seed),
+			fmt.Sprintf("k8sd-%d-worker-kubelet-client", seed),
+			fmt.Sprintf("k8sd-%d-worker-kube-proxy-client", seed),
+		},
+	}
+
+	err = os.WriteFile(filepath.Join(snap.KubernetesPKIDir(), "kubelet.key.new"), []byte(pKeyKubeletServing), 0600)
+	if err != nil {
+		return response.InternalError(err)
+	}
+	err = os.WriteFile(filepath.Join(snap.KubernetesPKIDir(), "kubelet-client.key.new"), []byte(pKeyKubeletClient), 0600)
+	if err != nil {
+		return response.InternalError(err)
+	}
+	err = os.WriteFile(filepath.Join(snap.KubernetesPKIDir(), "kube-proxy.key.new"), []byte(pKeyKubeProxy), 0600)
+	if err != nil {
+		return response.InternalError(err)
+	}
+	return response.SyncResponse(true, &result)
+
+}
+
+func (e *Endpoints) postRefreshCertsRun(s *state.State, r *http.Request) response.Response {
+	// TODO: Control Plane refresh
+	req := apiv1.RefreshCertificatesRunRequest{}
+	if err := utils.NewStrictJSONDecoder(r.Body).Decode(&req); err != nil {
+		return response.BadRequest(fmt.Errorf("failed to parse request: %w", err))
+	}
+	snap := e.provider.Snap()
+
+	client, err := snap.KubernetesNodeClient("")
+	if err != nil {
+		return response.InternalError(err)
+	}
+	csrNames := []string{
+		fmt.Sprintf("k8sd-%d-worker-kubelet-serving", req.Seed),
+		fmt.Sprintf("k8sd-%d-worker-kubelet-client", req.Seed),
+		fmt.Sprintf("k8sd-%d-worker-kube-proxy-client", req.Seed),
+	}
+
+	certificates := pki.WorkerNodePKI{}
+
+	for _, csrName := range csrNames {
+		csr, err := client.GetCertificateSigningRequest(s.Context, csrName)
+		if err != nil {
+			return response.InternalError(err)
+		}
+
+		if !isCertificateApproved(csr.Status.Conditions) {
+			return response.InternalError(fmt.Errorf("CSR %s is not issued", csrName))
```

Comment:
`s/issued/approved/`?
I'm always confused when a certificate is issued vs. approved :)

Reply:
Yes, if the CSR condition is approved, we know that the certificate has been approved using kubectl. Unfortunately, there's no condition for issuance, so the only way to verify if the certificate has been issued is to check the certificate field. I'll update the errors to make them clearer.

Reply:
More information: https://kubernetes.io/docs/reference/kubernetes-api/authentication-resources/certificate-signing-request-v1/#CertificateSigningRequestStatus

Reply:
I see, thanks for the context. 
I think from a usage perspective, something like this would probably be more readable:
```
func isCSRApproved(csr) bool
```
as it would hide the implementation details about when a CSR is approved.